mqtt/mqtt-bridge/src/pump/mod.rs

```diff
@@ -0,0 +1,334 @@
+                PumpMessage::ConfigurationUpdate(_) => {}
+                PumpMessage::Shutdown => {
+                    info!("stop requested");
+                    break;
+                }
+            }
+        }
+
+        info!("finished pump messages processor");
+    }
+}
+
+struct IngressShutdownHandle(Option<ClientShutdownHandle>);
+
+impl IngressShutdownHandle {
+    async fn shutdown(mut self) {
+        if let Some(mut sender) = self.0.take() {
+            if let Err(e) = sender.shutdown().await {
+                error!("unable to request shutdown for ingress. {}", e);
+            }
+        }
+    }
+}
+
+struct Ingress<H> {
+    client: MqttClient<H>,
+    shutdown_client: Option<ClientShutdownHandle>,
+}
+
+impl<H> Ingress<H>
+where
+    H: EventHandler,
+{
+    fn new(client: MqttClient<H>, shutdown_client: ClientShutdownHandle) -> Self {
+        Self {
+            client,
+            shutdown_client: Some(shutdown_client),
+        }
+    }
+
+    fn handle(&mut self) -> IngressShutdownHandle {
+        IngressShutdownHandle(self.shutdown_client.take())
+    }
+
+    async fn run(mut self) {
+        debug!("starting ingress publication processing",);
+        self.client.handle_events().await;
+    }
+}
+
+struct EgressShutdownHandle(Option<oneshot::Sender<()>>);
+
+impl EgressShutdownHandle {
+    async fn shutdown(mut self) {
+        if let Some(sender) = self.0.take() {
+            if sender.send(()).is_err() {
+                error!("unable to request shutdown for egress.");
+            }
+        }
+    }
+}
+
+struct Egress {
+    publish_handle: PublishHandle,
+    loader: MessageLoader<WakingMemoryStore>,
+    store: PublicationStore<WakingMemoryStore>,
+    shutdown_send: Option<oneshot::Sender<()>>,
+    shutdown_recv: oneshot::Receiver<()>,
+}
+
+impl Egress {
+    fn new(
+        publish_handle: PublishHandle,
+        loader: MessageLoader<WakingMemoryStore>,
+        store: PublicationStore<WakingMemoryStore>,
+    ) -> Egress {
+        let (shutdown_send, shutdown_recv) = oneshot::channel();
+
+        Self {
+            publish_handle,
+            loader,
+            store,
+            shutdown_send: Some(shutdown_send),
+            shutdown_recv,
+        }
+    }
+
+    fn handle(&mut self) -> EgressShutdownHandle {
+        EgressShutdownHandle(self.shutdown_send.take())
+    }
+
+    async fn run(self) {
+        let Egress {
+            mut publish_handle,
+            loader,
+            store,
+            shutdown_recv,
+            ..
+        } = self;
+
+        info!("starting egress publication processing");
+
+        let mut shutdown = shutdown_recv.fuse();
+        let mut loader = loader.fuse();
+
+        loop {
+            select! {
+                _ = &mut shutdown => {
+                    info!(" received shutdown signal for egress messages");
+                    break;
+                }
```

Comment:
I was thinking of using futures unordered collection to track all in flight messages so that they can be drained. I think we should discuss this approach.

Reply:
Yeah, lets discuss